frontend/src/metabase/dashboard/components/Dashboard/Dashboard.jsx

```diff
@@ -71,20 +67,43 @@ class DashboardInner extends Component {
       return;
     }
 
+    const isParameterValuesChanged = !_.isEqual(
+      prevProps.parameterValues,
+      this.props.parameterValues,
+    );
+    // This is a way of detecting card parameter mapping changes. Parameter
+    // mappings change happen independently of isEditing because of re-renders,
+    // so we use this heuristic to avoid additional state. The fetching logic
+    // should be moved outside.
+    const isDirtyChanged = !_.isEqual(
+      this._getDashcardsIsDirty(prevProps.dashboard?.dashcards),
+      this._getDashcardsIsDirty(this.props.dashboard?.dashcards),
+    );
+    const isDashboardLoaded = !prevProps.dashboard && this.props.dashboard;
+
     if (
-      !_.isEqual(prevProps.parameterValues, this.props.parameterValues) ||
-      (!prevProps.dashboard && this.props.dashboard)
+      isParameterValuesChanged ||
+      (!this.props.isEditing && isDirtyChanged) ||
+      isDashboardLoaded
     ) {
       this.props.fetchDashboardCardData({ reload: false, clearCache: true });
     }
   }
 
+  _getDashcardsIsDirty(dashcards = []) {
+    return dashcards.reduce((acc, dashcard) => {
+      acc[dashcard.id] = dashcard.isDirty ?? false;
+      return acc;
+    }, {});
+  }
```

Comment:
This isn't really a method but I haven't seen any function dangling around in component files, so I am putting it here. And I don't think adhoc functions like this one deserve to be called "utils" as they most likely won't be used twice.